exp/kubeadm-ignition/internal/ignition/s3backend_backend.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignition
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	ignTypes "github.com/coreos/ignition/config/v2_2/types"
+	"github.com/google/uuid"
+	"net/url"
+	"strings"
+	"time"
+)
+
+var (
+	baseIgnitionUri = map[string]string{
+		"v1.15.11": "ignition-config/k8s-v1.15.11.ign",
+		"v1.16.8":  "ignition-config/k8s-v1.16.8.ign",
+		"v1.17.4":  "ignition-config/k8s-v1.17.4.ign",
+	}
+)
+
+const (
+	KubernetesDefaultVersion = "v1.17.4"
+	IngitionSchemaVersion    = "2.2.0"
```

Comment:
Ignition config spec 2.x is unmaintained and unsupported upstream.  New code should use 3.x, either with or without compatibility code for 2.x.  One approach is to generate 3.x structs and then use [ign-converter](https://github.com/coreos/ign-converter/) to downconvert when needed.

Reply:
just to add here: ign-converter also goes the other way, i.e. we're using it to migrate clusters provisioned with Ign0.x/spec 2 to Ign2.x/spec 3